blockly-rtc/src/WorkspaceClient.js

```diff
@@ -108,4 +109,93 @@ export default class WorkspaceClient {
         this.writeInProgress = false;
     };
 
+    /**
+     * Trigger an API call to query events from the database.
+     * @returns {!Array<!Object>} The result of processQueryResults_() or an
+     * empty array if the API call fails.
+     * @public
+     */
+    async queryDatabase() {
+      try {
+        const rows = await getEvents(this.lastSync);
+        return this.processQueryResults_(rows);
+      } catch {
+        return [];
+      };
+    };
+
+    /**
+     * Compare the order of events in the rows retrieved from the database to
+     * the stacks of local-only changes and provide a series of steps that
+     * will allow the server and local workspace to converge.
+     * @param {<!Array<!Object>} rows Rows of event entries retrieved by
+     * querying the database.
+     * @returns {!Array<!Object>} eventQueue An array of events and the
+     * direction they should be run.
+     * @private
+     */
+    processQueryResults_(rows) {
+        const eventQueue = [];
+
+        if (rows.length == 0) {
+            return eventQueue;
+        };
+    
+        this.lastSync = rows[rows.length-1].serverId;
+    
+        // No local changes.
+        if (this.notSent.length == 0 && this.inProgress.length == 0) {
+            rows.forEach((row) => {
+                eventQueue.push({
+                    event: JSON.parse(row.event),
+                    forward: true
+                });
+            });
+            return eventQueue;
+        };
+    
+        // Common root, remove common events from server events.
+        if (this.inProgress.length > 0 && rows[0].entryId == this.inProgress[0].entryId) {
+            rows = rows.slice(this.inProgress.length);
+            this.inProgress = [];
+        };
+    
+        if (rows.length > 0) {
+            // Undo local events.
+            this.notSent.slice().reverse().forEach((row) => {
+                eventQueue.push({
+                    event: row.event,
```

Comment:
Sometimes you're calling `JSON.parse(row.event)` and not others.  What's the difference?

Reply:
And if they should all be one way or the other, you can pull the callback out into a separate function since it's the same for each of these `forEach` calls.

Reply:
The difference is that the events are stored differently. There is actually a little more background on this that might be worth discussing offline if you have a chance.